File: Velix/app/lib/services/chat_service.dart
import 'dart:async';
import 'dart:convert';
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_auth/firebase_auth.dart';
import 'package:flutter/foundation.dart';
import '../models/user_info.dart';
import '../models/chat_message.dart';
import 'database_service.dart';
import 'encryption_service.dart';
import 'media_service.dart';

class _PendingDelete {
  final DocumentReference ref;
  final String? cloudinaryPublicId;
  final String? mediaType;
  _PendingDelete(this.ref, {this.cloudinaryPublicId, this.mediaType});
}

class ChatService {
  final EncryptionService _encryptionService;
  final _mediaService = MediaService();
  final FirebaseFirestore _firestore = FirebaseFirestore.instance;

  User? _currentUser;
  String? _currentHandle;
  StreamSubscription? _messagesSub;
  String? _listeningPeerId;
  bool _disposed = false;

  final List<_PendingDelete> _pendingDeletes = [];
  final Map<String, String> _peerPublicKeys = {};

  final _messageController = StreamController<ChatMessage>.broadcast();
  final _eventController = StreamController<Map<String, dynamic>>.broadcast();

  ChatService(this._encryptionService);

  String? get currentUserId => _currentUser?.uid;
  String? get currentHandle => _currentHandle;
  Stream<ChatMessage> get messageStream => _messageController.stream;
  Stream<Map<String, dynamic>> get eventStream => _eventController.stream;
  Stream<Map<String, dynamic>> get burnAckStream => _eventController.stream;

  Future<void> initialize(User user) async {
    _currentUser = user;
    await _encryptionService.generateKeyPair();

    // Load existing doc to preserve userHandle
    final docSnap =
        await _firestore.collection('users').doc(user.uid).get();
    _currentHandle = docSnap.data()?['userHandle'] as String?;

    // If no handle yet, generate a default from UID
    if (_currentHandle == null || _currentHandle!.isEmpty) {
      final raw = user.uid.replaceAll(RegExp(r'[^a-z0-9]'), '');
      _currentHandle = 'u_${raw.substring(0, raw.length.clamp(0, 8))}';
    }

    await _firestore.collection('users').doc(user.uid).set({
      'displayName': user.displayName ?? user.email ?? 'Unknown',
      'photoURL': user.photoURL ?? '',
      'publicKey': _encryptionService.publicKeyBase64,
      'userHandle': _currentHandle,
      'lastSeen': FieldValue.serverTimestamp(),
    }, SetOptions(merge: true));

    // Cache own user data locally in SQLite
    await DatabaseService.instance.upsertUser(
      uid: user.uid,
      handle: _currentHandle,
      name: user.displayName ?? user.email ?? 'Unknown',
      photoUrl: user.photoURL ?? '',
      publicKey: _encryptionService.publicKeyBase64,
    );
  }

  // Returns null on success, error message string on failure.
  Future<String?> updateHandle(String newHandle) async {
    final handle = newHandle.toLowerCase().trim();
    if (handle.length < 3) return '用戶 ID 至少需要 3 個字元';
    if (handle.length > 20) return '用戶 ID 最多 20 個字元';
    if (!RegExp(r'^[a-z0-9_]+$').hasMatch(handle)) {
      return '只能使用英文小寫字母、數字與底線 (_)';
    }

    final snap = await _firestore
        .collection('users')
        .where('userHandle', isEqualTo: handle)
        .limit(1)
        .get();

    final taken =
        snap.docs.isNotEmpty && snap.docs.first.id != _currentUser!.uid;
    if (taken) return '此用戶 ID 已被使用，請換一個';

    await _firestore
        .collection('users')
        .doc(_currentUser!.uid)
        .update({'userHandle': handle});
    _currentHandle = handle;
    return null;
  }

  // Search a user by exact handle. Returns null if not found.
  Future<ChatUser?> searchUser(String handle) async {
    final snap = await _firestore
        .collection('users')
        .where('userHandle', isEqualTo: handle.toLowerCase().trim())
        .limit(1)
        .get();

    if (snap.docs.isEmpty) return null;
    final d = snap.docs.first;
    if (d.id == _currentUser?.uid) return null; // can't chat with yourself
    final data = d.data();
    final pk = data['publicKey'] as String? ?? '';
    _peerPublicKeys[d.id] = pk;

    // Cache peer data in local SQLite
    await DatabaseService.instance.upsertUser(
      uid: d.id,
      handle: data['userHandle'] as String?,
      name: data['displayName'] as String?,
      photoUrl: data['photoURL'] as String?,
      publicKey: pk,
    );

    return ChatUser(
      userId: d.id,
      displayName: data['displayName'] as String? ?? d.id,
      photoURL: data['photoURL'] as String? ?? '',
      publicKey: pk,
      userHandle: data['userHandle'] as String? ?? '',
    );
  }

  String _chatId(String peerId) {
    final ids = [_currentUser!.uid, peerId]..sort();
    return '${ids[0]}_${ids[1]}';
  }

  void listenToMessages(String peerId) {
    if (_listeningPeerId == peerId) return;
    _listeningPeerId = peerId;
    _messagesSub?.cancel();
    _pendingDeletes.clear();

    _messagesSub = _firestore
        .collection('chats')
        .doc(_chatId(peerId))
        .collection('messages')
        .orderBy('timestamp')
        .snapshots()
        .listen((snap) {
      for (final change in snap.docChanges) {
        switch (change.type) {
          case DocumentChangeType.added:
            _processIncoming(change.doc, peerId);
          case DocumentChangeType.modified:
            _processModified(change.doc, peerId);
          case DocumentChangeType.removed:
            _handleRemoved(change.doc);
        }
      }
    });
  }

  Future<void> _processIncoming(
    DocumentSnapshot<Map<String, dynamic>> doc,
    String peerId,
  ) async {
    try {
      final data = doc.data()!;
      final from = data['from'] as String;
      if (from == _currentUser!.uid) return;

      // Always fetch fresh public key from Firestore — peer may have restarted.
      // forceRefresh bypasses both in-memory and SQLite caches.
      _peerPublicKeys.remove(from);
      _encryptionService.clearSharedSecret(from);

      final burnTimer = data['burnTimer'] as String? ?? 'off';
      final cloudinaryPublicId = data['cloudinaryPublicId'] as String?;
      final rawMediaType = data['mediaType'] as String?;

      if (data['recalled'] == true) {
        if (!_messageController.isClosed) {
          _messageController.add(ChatMessage(
            documentId: doc.id,
            from: from,
            to: _currentUser!.uid,
            text: '',
            timestamp:
                (data['timestamp'] as Timestamp?)?.toDate() ?? DateTime.now(),
            burnTimer: burnTimer,
            recalled: true,
            isSentByMe: false,
          ));
        }
        doc.reference.delete().catchError((_) {});
        return;
      }

      final publicKey = await _resolvePublicKey(from, forceRefresh: true);
      if (publicKey.isEmpty) {
        debugPrint('[ChatService] publicKey empty for $from — skipping');
        return;
      }

      final sharedSecret =
          await _encryptionService.getSharedSecret(from, publicKey);
      final plainText =
          await _encryptionService.decryptFromMap(data, sharedSecret);
      final payload = jsonDecode(plainText) as Map<String, dynamic>;

      debugPrint('[ChatService] recv from=$from mediaType=${payload['mediaType']} hasUrl=${payload['mediaUrl'] != null}');

      if (!_messageController.isClosed) {
        _messageController.add(ChatMessage(
          documentId: doc.id,
          from: from,
          to: _currentUser!.uid,
          text: payload['text'] as String? ?? '',
          mediaUrl: payload['mediaUrl'] as String?,
          mediaType: payload['mediaType'] as String?,
          fileName: payload['fileName'] as String?,
          timestamp:
              (data['timestamp'] as Timestamp?)?.toDate() ?? DateTime.now(),
          burnTimer: burnTimer,
          edited: data['edited'] as bool? ?? false,
          isSentByMe: false,
        ));
      }

      switch (burnTimer) {
        case '1m':
          Future.delayed(const Duration(minutes: 1), () =>
              _burnCloudinaryAndDoc(
                  doc.reference, cloudinaryPublicId, rawMediaType));
        case '3m':
          Future.delayed(const Duration(minutes: 3), () =>
              _burnCloudinaryAndDoc(
                  doc.reference, cloudinaryPublicId, rawMediaType));
        case '5m':
          Future.delayed(const Duration(minutes: 5), () =>
              _burnCloudinaryAndDoc(
                  doc.reference, cloudinaryPublicId, rawMediaType));
        default:
          _pendingDeletes.add(_PendingDelete(
            doc.reference,
            cloudinaryPublicId: cloudinaryPublicId,
            mediaType: rawMediaType,
          ));
      }
    } catch (e, st) {
      debugPrint('[ChatService] _processIncoming error: $e');
      debugPrint('[ChatService] stack: $st');
    }
  }

  void _burnCloudinaryAndDoc(
    DocumentReference ref,
    String? publicId,
    String? mediaType,
  ) {
    if (publicId != null && mediaType != null) {
      final resourceType = mediaType == 'file' ? 'raw' : mediaType;
      _mediaService.delete(publicId, resourceType);
    }
    ref.delete().catchError((_) {});
  }

  Future<void> _processModified(
    DocumentSnapshot<Map<String, dynamic>> doc,
    String peerId,
  ) async {
    try {
      final data = doc.data()!;
      final from = data['from'] as String;
      if (from == _currentUser!.uid) return;

      if (data['recalled'] == true) {
        if (!_eventController.isClosed) {
          _eventController.add({'type': 'recall', 'documentId': doc.id});
        }
        doc.reference.delete().catchError((_) {});
        return;
      }

      if (data['edited'] == true) {
        final publicKey = await _resolvePublicKey(from);
        if (publicKey.isEmpty) return;
        final sharedSecret =
            await _encryptionService.getSharedSecret(from, publicKey);
        final plainText =
            await _encryptionService.decryptFromMap(data, sharedSecret);
        final payload = jsonDecode(plainText) as Map<String, dynamic>;
        if (!_eventController.isClosed) {
          _eventController.add({
            'type': 'edit',
            'documentId': doc.id,
            'newText': payload['text'] as String,
          });
        }
      }
    } catch (_) {}
  }

  void _handleRemoved(DocumentSnapshot<Map<String, dynamic>> doc) {
    final data = doc.data();
    if (data == null) return;
    final burnTimer = data['burnTimer'] as String? ?? 'off';
    final msgFrom = data['from'] as String?;

    if (burnTimer != 'off' && msgFrom == _currentUser!.uid) {
      if (!_eventController.isClosed) {
        _eventController.add({
          'type': 'burn',
          'documentId': doc.id,
          'from': data['to'] as String? ?? '',
        });
      }
    }
  }

  // forceRefresh=true skips all caches and always fetches from Firestore.
  // Used by _processIncoming so we always decrypt with the peer's latest key.
  Future<String> _resolvePublicKey(String uid, {bool forceRefresh = false}) async {
    if (!forceRefresh) {
      // 1. In-memory cache
      if (_peerPublicKeys[uid]?.isNotEmpty == true) return _peerPublicKeys[uid]!;

      // 2. Local SQLite cache
      final cached = await DatabaseService.instance.getUser(uid);
      final cachedPk = cached?['public_key'] as String?;
      if (cachedPk != null && cachedPk.isNotEmpty) {
        _peerPublicKeys[uid] = cachedPk;
        return cachedPk;
      }
    }

    // 3. Firestore — always reached when forceRefresh=true
    final doc = await _firestore.collection('users').doc(uid).get();
    final pk = doc.data()?['publicKey'] as String? ?? '';
    _peerPublicKeys[uid] = pk;
    if (pk.isNotEmpty) {
      await DatabaseService.instance.upsertUser(
        uid: uid,
        publicKey: pk,
        handle: doc.data()?['userHandle'] as String?,
        name: doc.data()?['displayName'] as String?,
        photoUrl: doc.data()?['photoURL'] as String?,
      );
    }
    return pk;
  }

  Future<void> sendMessage(
    String peerId,
    String text, {
    String burnTimer = 'off',
    String? mediaUrl,
    String? mediaType,
    String? fileName,
    String? cloudinaryPublicId,
  }) async {
    final publicKey = await _resolvePublicKey(peerId);
    if (publicKey.isEmpty) return;

    final sharedSecret =
        await _encryptionService.getSharedSecret(peerId, publicKey);
    final plainPayload = jsonEncode({
      'text': text,
      'mediaUrl': mediaUrl,
      'mediaType': mediaType,
      'fileName': fileName,
      'timestamp': DateTime.now().millisecondsSinceEpoch,
    });
    final encrypted =
        await _encryptionService.encryptToMap(plainPayload, sharedSecret);

    final docRef = await _firestore
        .collection('chats')
        .doc(_chatId(peerId))
        .collection('messages')
        .add({
      'from': _currentUser!.uid,
      'to': peerId,
      ...encrypted,
      'burnTimer': burnTimer,
      'mediaType': mediaType,
      'cloudinaryPublicId': cloudinaryPublicId,
      'fileName': fileName,
      'recalled': false,
      'edited': false,
      'timestamp': FieldValue.serverTimestamp(),
    });

    if (!_messageController.isClosed) {
      _messageController.add(ChatMessage(
        documentId: docRef.id,
        from: _currentUser!.uid,
        to: peerId,
        text: text,
        mediaUrl: mediaUrl,
        mediaType: mediaType,
        fileName: fileName,
        timestamp: DateTime.now(),
        burnTimer: burnTimer,
        isSentByMe: true,
      ));
    }
  }

  Future<void> editMessage(
      String documentId, String peerId, String newText) async {
    final publicKey = await _resolvePublicKey(peerId);
    if (publicKey.isEmpty) return;

    final sharedSecret =
        await _encryptionService.getSharedSecret(peerId, publicKey);
    final plainPayload = jsonEncode({
      'text': newText,
      'timestamp': DateTime.now().millisecondsSinceEpoch,
    });
    final encrypted =
        await _encryptionService.encryptToMap(plainPayload, sharedSecret);

    await _firestore
        .collection('chats')
        .doc(_chatId(peerId))
        .collection('messages')
        .doc(documentId)
        .update({...encrypted, 'edited': true});

    if (!_eventController.isClosed) {
      _eventController.add({
        'type': 'edit',
        'documentId': documentId,
        'newText': newText,
      });
    }
  }

  Future<void> recallMessage(String documentId, String peerId) async {
    await _firestore
        .collection('chats')
        .doc(_chatId(peerId))
        .collection('messages')
        .doc(documentId)
        .update({'recalled': true, 'ct': '', 'nonce': '', 'mac': ''});

    if (!_eventController.isClosed) {
      _eventController.add({'type': 'recall', 'documentId': documentId});
    }
  }

  void stopListening() {
    for (final pending in _pendingDeletes) {
      _burnCloudinaryAndDoc(
        pending.ref,
        pending.cloudinaryPublicId,
        pending.mediaType,
      );
    }
    _pendingDeletes.clear();
    _messagesSub?.cancel();
    _listeningPeerId = null;
  }

  Future<void> burnMessage(String documentId, String peerId) async {}

  void dispose() {
    if (_disposed) return;
    _disposed = true;
    stopListening();
    _messageController.close();
    _eventController.close();
  }
}

File: P-Chats/frontend/src/services/chatService.ts
import {
  collection,
  doc,
  addDoc,
  updateDoc,
  getDoc,
  onSnapshot,
  orderBy,
  query,
  serverTimestamp,
  where,
  limit,
  getDocs,
  type Unsubscribe,
} from 'firebase/firestore'
import { db } from '../lib/firebase'
import { EncryptionService } from '../lib/encryption'

export interface ChatMessage {
  documentId: string
  from: string
  to: string
  text: string
  timestamp: Date
  burnTimer: 'off' | 'exit' | '1m' | '3m' | '5m'
  mediaUrl?: string
  mediaType?: 'image' | 'video' | 'file'
  fileName?: string
  isBurned: boolean
  isSentByMe: boolean
  recalled: boolean
  edited: boolean
}

export interface ChatUser {
  userId: string
  displayName: string
  photoURL: string
  publicKey: string
  userHandle: string
}

interface PendingDelete {
  docPath: string
  cloudinaryPublicId?: string
  mediaType?: string
}

type MessageHandler = (msg: ChatMessage) => void
type EventHandler = (event: { type: string; documentId: string; newText?: string }) => void

export class ChatService {
  private enc: EncryptionService
  private currentUserId: string
  private currentHandle: string
  private currentDisplayName: string
  private peerPublicKeys = new Map<string, string>()
  private pendingDeletes: PendingDelete[] = []
  private unsub: Unsubscribe | null = null
  private listeningPeerId: string | null = null
  private onMessage: MessageHandler = () => {}
  private onEvent: EventHandler = () => {}

  constructor(
    enc: EncryptionService,
    userId: string,
    handle: string,
    displayName: string,
  ) {
    this.enc = enc
    this.currentUserId = userId
    this.currentHandle = handle
    this.currentDisplayName = displayName
  }

  get handle() { return this.currentHandle }
  get displayName() { return this.currentDisplayName }
  get userId() { return this.currentUserId }

  setHandlers(onMessage: MessageHandler, onEvent: EventHandler) {
    this.onMessage = onMessage
    this.onEvent = onEvent
  }

  private chatId(peerId: string): string {
    return [this.currentUserId, peerId].sort().join('_')
  }

  async updateHandle(newHandle: string): Promise<string | null> {
    const h = newHandle.toLowerCase().trim()
    if (h.length < 3) return '用戶 ID 至少需要 3 個字元'
    if (h.length > 20) return '用戶 ID 最多 20 個字元'
    if (!/^[a-z0-9_]+$/.test(h)) return '只能使用英文小寫字母、數字與底線 (_)'
    const snap = await getDocs(
      query(collection(db, 'users'), where('userHandle', '==', h), limit(1)),
    )
    if (!snap.empty && snap.docs[0].id !== this.currentUserId) return '此用戶 ID 已被使用，請換一個'
    await updateDoc(doc(db, 'users', this.currentUserId), { userHandle: h })
    this.currentHandle = h
    return null
  }

  async updateDisplayName(newName: string): Promise<string | null> {
    const name = newName.trim()
    if (!name) return '顯示名稱不得為空'
    if (name.length > 30) return '顯示名稱最多 30 個字元'
    await updateDoc(doc(db, 'users', this.currentUserId), { displayName: name })
    this.currentDisplayName = name
    return null
  }

  async deleteUserData(): Promise<void> {
    const { deleteDoc: del } = await import('firebase/firestore')
    await del(doc(db, 'users', this.currentUserId)).catch(() => {})
  }

  async searchUser(handle: string): Promise<ChatUser | null> {
    const snap = await getDocs(
      query(collection(db, 'users'), where('userHandle', '==', handle.toLowerCase().trim()), limit(1)),
    )
    if (snap.empty) return null
    const d = snap.docs[0]
    if (d.id === this.currentUserId) return null
    const data = d.data()
    this.peerPublicKeys.set(d.id, data.publicKey ?? '')
    return {
      userId: d.id,
      displayName: data.displayName ?? d.id,
      photoURL: data.photoURL ?? '',
      publicKey: data.publicKey ?? '',
      userHandle: data.userHandle ?? '',
    }
  }

  private async resolvePublicKey(uid: string, forceRefresh = false): Promise<string> {
    if (!forceRefresh) {
      const cached = this.peerPublicKeys.get(uid)
      if (cached) return cached
    }
    const snap = await getDoc(doc(db, 'users', uid))
    const pk = (snap.data()?.publicKey as string) ?? ''
    this.peerPublicKeys.set(uid, pk)
    return pk
  }

  listenToMessages(peerId: string) {
    if (this.listeningPeerId === peerId) return
    this.stopListening()
    this.listeningPeerId = peerId

    const messagesRef = collection(db, 'chats', this.chatId(peerId), 'messages')
    const q = query(messagesRef, orderBy('timestamp'))

    this.unsub = onSnapshot(q, snap => {
      snap.docChanges().forEach(change => {
        if (change.type === 'added') this.processIncoming(change.doc, peerId)
        else if (change.type === 'modified') this.processModified(change.doc, peerId)
        else if (change.type === 'removed') this.handleRemoved(change.doc)
      })
    })
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private async processIncoming(docSnap: any, peerId: string): Promise<void> {
    try {
      const data = docSnap.data() as Record<string, unknown>
      const from = data.from as string
      if (from === this.currentUserId) return

      this.peerPublicKeys.delete(from)
      this.enc.clearSharedSecret(from)

      const burnTimer = (data.burnTimer as string) ?? 'off'
      const cloudinaryPublicId = data.cloudinaryPublicId as string | undefined
      const rawMediaType = data.mediaType as string | undefined

      if (data.recalled === true) {
        this.onMessage({
          documentId: docSnap.id,
          from, to: this.currentUserId,
          text: '', timestamp: (data.timestamp as { toDate(): Date })?.toDate() ?? new Date(),
          burnTimer: burnTimer as ChatMessage['burnTimer'],
          isBurned: false, isSentByMe: false, recalled: true, edited: false,
        })
        docSnap.ref.delete().catch(() => {})
        return
      }

      const pk = await this.resolvePublicKey(from, true)
      if (!pk) return
      const sharedKey = await this.enc.getSharedSecret(from, pk)
      const plainText = await this.enc.decryptFromMap(
        { ct: data.ct as string, nonce: data.nonce as string, mac: data.mac as string },
        sharedKey,
      )
      const payload = JSON.parse(plainText) as Record<string, unknown>

      this.onMessage({
        documentId: docSnap.id,
        from, to: this.currentUserId,
        text: (payload.text as string) ?? '',
        mediaUrl: payload.mediaUrl as string | undefined,
        mediaType: payload.mediaType as ChatMessage['mediaType'],
        fileName: payload.fileName as string | undefined,
        timestamp: (data.timestamp as { toDate(): Date })?.toDate() ?? new Date(),
        burnTimer: burnTimer as ChatMessage['burnTimer'],
        edited: (data.edited as boolean) ?? false,
        isBurned: false, isSentByMe: false, recalled: false,
      })

      const delay = burnTimer === '1m' ? 60_000 : burnTimer === '3m' ? 180_000 : burnTimer === '5m' ? 300_000 : null
      if (delay !== null) {
        setTimeout(() => {
          if (cloudinaryPublicId && rawMediaType) {
            this.deleteCloudinary(cloudinaryPublicId, rawMediaType === 'file' ? 'raw' : rawMediaType)
          }
          docSnap.ref.delete().catch(() => {})
        }, delay)
      } else {
        this.pendingDeletes.push({ docPath: docSnap.ref.path, cloudinaryPublicId, mediaType: rawMediaType })
      }
    } catch (e) {
      console.error('[ChatService] processIncoming error', e)
    }
  }

  private async processModified(docSnap: any, _peerId: string): Promise<void> {
    try {
      const data = docSnap.data() as Record<string, unknown>
      const from = data.from as string
      if (from === this.currentUserId) return

      if (data.recalled === true) {
        this.onEvent({ type: 'recall', documentId: docSnap.id })
        docSnap.ref.delete().catch(() => {})
        return
      }
      if (data.edited === true) {
        const pk = await this.resolvePublicKey(from)
        if (!pk) return
        const sharedKey = await this.enc.getSharedSecret(from, pk)
        const plainText = await this.enc.decryptFromMap(
          { ct: data.ct as string, nonce: data.nonce as string, mac: data.mac as string },
          sharedKey,
        )
        const payload = JSON.parse(plainText) as Record<string, unknown>
        this.onEvent({ type: 'edit', documentId: docSnap.id, newText: payload.text as string })
      }
    } catch {}
  }

  private handleRemoved(docSnap: any): void {
    const data = docSnap.data() as Record<string, unknown> | undefined
    if (!data) return
    const burnTimer = (data.burnTimer as string) ?? 'off'
    if (burnTimer !== 'off' && data.from === this.currentUserId) {
      this.onEvent({ type: 'burn', documentId: docSnap.id })
    }
  }

  async sendMessage(
    peerId: string,
    text: string,
    opts: {
      burnTimer?: ChatMessage['burnTimer']
      mediaUrl?: string
      mediaType?: ChatMessage['mediaType']
      fileName?: string
      cloudinaryPublicId?: string
    } = {},
  ): Promise<void> {
    const pk = await this.resolvePublicKey(peerId)
    if (!pk) return
    const sharedKey = await this.enc.getSharedSecret(peerId, pk)
    const plainPayload = JSON.stringify({
      text,
      mediaUrl: opts.mediaUrl,
      mediaType: opts.mediaType,
      fileName: opts.fileName,
      timestamp: Date.now(),
    })
    const encrypted = await this.enc.encryptToMap(plainPayload, sharedKey)
    const docRef = await addDoc(collection(db, 'chats', this.chatId(peerId), 'messages'), {
      from: this.currentUserId,
      to: peerId,
      ...encrypted,
      burnTimer: opts.burnTimer ?? 'off',
      mediaType: opts.mediaType ?? null,
      cloudinaryPublicId: opts.cloudinaryPublicId ?? null,
      fileName: opts.fileName ?? null,
      recalled: false,
      edited: false,
      timestamp: serverTimestamp(),
    })

    this.onMessage({
      documentId: docRef.id,
      from: this.currentUserId,
      to: peerId,
      text,
      mediaUrl: opts.mediaUrl,
      mediaType: opts.mediaType,
      fileName: opts.fileName,
      timestamp: new Date(),
      burnTimer: opts.burnTimer ?? 'off',
      isBurned: false,
      isSentByMe: true,
      recalled: false,
      edited: false,
    })
  }

  async editMessage(documentId: string, peerId: string, newText: string): Promise<void> {
    const pk = await this.resolvePublicKey(peerId)
    if (!pk) return
    const sharedKey = await this.enc.getSharedSecret(peerId, pk)
    const plainPayload = JSON.stringify({ text: newText, timestamp: Date.now() })
    const encrypted = await this.enc.encryptToMap(plainPayload, sharedKey)
    await updateDoc(doc(db, 'chats', this.chatId(peerId), 'messages', documentId), {
      ...encrypted,
      edited: true,
    })
    this.onEvent({ type: 'edit', documentId, newText })
  }

  async recallMessage(documentId: string, peerId: string): Promise<void> {
    await updateDoc(doc(db, 'chats', this.chatId(peerId), 'messages', documentId), {
      recalled: true, ct: '', nonce: '', mac: '',
    })
    this.onEvent({ type: 'recall', documentId })
  }

  async burnMessage(documentId: string, peerId: string): Promise<void> {
    const { deleteDoc: del } = await import('firebase/firestore')
    await del(doc(db, 'chats', this.chatId(peerId), 'messages', documentId)).catch(() => {})
  }

  private deleteCloudinary(publicId: string, resourceType: string): void {
    // Cloudinary deletion requires a signed request; done via the media service
    import('../services/mediaService').then(({ mediaService }) => {
      mediaService.delete(publicId, resourceType)
    })
  }

  stopListening(): void {
    for (const pending of this.pendingDeletes) {
      if (pending.cloudinaryPublicId && pending.mediaType) {
        const rt = pending.mediaType === 'file' ? 'raw' : pending.mediaType
        this.deleteCloudinary(pending.cloudinaryPublicId, rt)
      }
      import('firebase/firestore').then(({ doc: d, deleteDoc: del }) => {
        const ref = d(db, pending.docPath)
        del(ref).catch(() => {})
      })
    }
    this.pendingDeletes = []
    this.unsub?.()
    this.unsub = null
    this.listeningPeerId = null
  }

  dispose(): void {
    this.stopListening()
  }
}
